### PR TITLE
WIP: Undo / Redo functionality

### PR DESCRIPTION
**Pull request by harshad1**
Attempting to understand and fix https://github.com/gsantner/markor/issues/894.

This PR is nowhere close to ready. I am unable to understand how certain parts of the code work and so I am creating this so I can ask @gsantner :)

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java [hunk 1]
```diff
@@ -180,8 +177,9 @@ public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
     }
 
     @Override
-    public void onResume() {
-        super.onResume();
+    public void onStart() {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:180]: Codacy Issue found: [The method 'onStart()' has an NPath complexity of 500](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905&bid=17562240)

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java [hunk 2]
```diff
@@ -469,21 +473,7 @@ public void onFsViewerConfig(FilesystemViewerData.Options dopt) {
...
-                if (activity != null && activity instanceof AppCompatActivity) {
-                    ((AppCompatActivity) activity).supportInvalidateOptionsMenu();
-                }
-            }, HISTORY_DELTA);
-        }
-        Activity activity = getActivity();
-        if (activity != null && activity instanceof AppCompatActivity) {
-            ((AppCompatActivity) activity).supportInvalidateOptionsMenu();
-        }
-
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:486]: It seems that the options menu for the current activity is invalidated immediately and then again after 5 seconds ? @gsantner Could you help me understand what is being done here?

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905&bid=17562240)

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java [hunk 3]
```diff
@@ -182,6 +179,13 @@ public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
     @Override
     public void onResume() {
         super.onResume();
+        onPrepareOptionsMenu(getFragmentMenu());
+    }
+
+    @Override
+    public void onStart() {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:186]: Codacy Issue found: [The method 'onStart()' has an NPath complexity of 500](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905&bid=17562240)

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java [hunk 4]
```diff
@@ -182,42 +179,62 @@ public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
...
+            _textSdWarning.setVisibility(permok ? View.GONE : View.VISIBLE);
+
+            if (_shareUtil.isUnderStorageAccessFolder(file) && _shareUtil.getStorageAccessFrameworkTreeUri() == null) {
                 _shareUtil.showMountSdDialog(getActivity());
                 return;
             }
-            _textSdWarning.setVisibility(permok ? View.GONE : View.VISIBLE);
+
+            if (getActivity() instanceof DocumentActivity) {
+                if (file.getAbsolutePath().contains("mordor/1-epub-experiment.md")) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:206]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java [hunk 5]
```diff
@@ -73,6 +74,7 @@ protected static Highlighter getDefaultHighlighter(HighlightingEditor hlEditor,
     protected boolean _highlightLinks = true;
     protected final boolean _highlightHexcolor;
     protected final Document _document;
+    private TextWatcher _modifier = null;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5359905&bid=17562240)

**Comment by gsantner** [comment]
this PR is deprecated in favor of #905 and can be closed right?

**Comment by harshad1** [comment]
Closing in favour of #905

**Pull request closed without merging**